### ISSUE TITLE
What is the recommended way to set flags in a CMake based project?

### ISSUE THREAD

**Issue by AliAskari** [conan-io/conan#4754]
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [N/A] I've specified the Conan version, operating system version and any tool that can be relevant.
- [N/A] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

I am trying to start a new CMake based project using Conan and I am confused how to set the flags. I am aware that the flags (listed [here](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#cpp-info)) can be specified in the Conan recipe. However these flags are only used when the project/library is built with Conan. If I want to build the project manually, without conan I am forced to specify the flags in a cmake file. Which forces me to always update both files if I need to change a flag. My first question is what is the recommended way for cases like this? Is there a workflow for cases like this?

Also, I believe it might have been more useful for the C++ community if there was an attempt to standardize the way setting flags work in CMake. Possibly a pull request to the original CMake project with a new cmake module that could take care of the flags in a standardized way. Something that would also work with Conan; such that the package maintainers would not have to take care of updating two different set of flags.

**Comment by lasote** [comment]
Hi, thanks for your question.
It depends, what kind of flags are you adjusting?

**Comment by AliAskari** [comment]
Other than the generic set of flags that I would like to use for all builds, `-Wall -Werror` et al, my main concern is how to integrate the sanitizer flags. There are at least two set of flags that I am interested in `-fsanitize=address -fno-omit-frame-pointer` and `-fsanitize=thread`. This means I would like to have 3 different binaries for each configuration (arch/build_type combinations):

1. Not sanitized at all and built with general flags that I have for all builds
2. Flags mentioned in 1. + `fsanitize=address -fno-omit-frame-pointer`
3. Flags mentioned in 1. + `fsanitize=thread`

I am not sure why you're saying it depends. Should it? Would it not be easier to generalize this for all flags? debug/release, 32/64 bit build, MT vs MD, these are all configurations that depend on the flag but just because they're used more often they're under special attention and have been abstracted away and are only considered as settings for a conan package.

**Comment by memsharded** [comment]
> Also, I believe it might have been more useful for the C++ community if there was an attempt to standardize the way setting flags work in CMake. Possibly a pull request to the original CMake project with a new cmake module that could take care of the flags in a standardized way. Something that would also work with Conan; such that the package maintainers would not have to take care of updating two different set of flags.

Yes, that is part of the problem, I'd also like to see this more standardized across build systems and compilers. 

If you want Conan to manage different binaries for sanitized/non-sanitized builds, then it is necessary to add this to the model. The model extension points are 2: settings and options. Settings are project-wide, and probably apply to sanitizers, as the whole project should be built with the same sanitizing configuration

 ``cmake.definitions["CONAN_CXX_FLAGS"] = os.getenv("MY_CXX_SANITIZER_FLAGS")``

For extending the settings, the steps would be something like:
- Add the settings you want to ``settings.yml`` in your local cache (remember this file can be shared with ``conan config install``)
- Allow a value of ``None`` to this new setting. Something like ``sanitizer = [None, address, thread]``
- Add the ``sanitizer`` setting to your recipes: ``settings = "os", "compiler", ..., "sanitizer"
- If you put sanitizer as a subsetting of the compiler, as probably they will be only defined for gcc or clang, then you don't need to add ``sanitizer`` to the recipe setting, as it will be ``self.settings.compiler.sanitizer``
- Use the ``sanitizer`` in your ``build()`` method, something like:

```python
def build(self):
     cmake = CMake(self):
     if self.settings.compiler.sanitizer == "address":
           cmake.definitions["CONAN_CXX_FLAGS"] = "-fsanitize=address -fno-omit-frame-pointer"
     elif self.settings.compiler.sanitizer == "thread":
           ...
```

**Comment by AliAskari** [comment]
I see, thanks for the detailed reply; it's clear now. However it appears that I still need to add the flags in my CMakefile in case I build the library without Conan, just pure CMake. I was looking for a way to set the flags only once, maybe in a file `compile_flags.cxx`, and then have CMake/Conan use that file to set the flags.

**Comment by lasote** [comment]
I understand. We have a pending feature of refactoring the build helpers, that way when you first run `conan build --configure` you might obtaining a file generated by your `CMake` build helper, containing all the adjusted flags, including your own `sanitize` ones. Once you have that file (different from the one for your dependencies that is the `conanbuildinfo.cmake` one) you can include it in your cmakelist os use it as a toolchain calling directly to `CMake`. But this is only the theory behind the roadmap feature.

**Comment by SeanSnyders** [comment]
> > Also, I believe it might have been more useful for the C++ community if there was an attempt to standardize the way setting flags work in CMake. Possibly a pull request to the original CMake project with a new cmake module that could take care of the flags in a standardized way. Something that would also work with Conan; such that the package maintainers would not have to take care of updating two different set of flags.
> 
> Yes, that is part of the problem, I'd also like to see this more standardized across build systems and compilers.
> 
> If you want Conan to manage different binaries for sanitized/non-sanitized builds, then it is necessary to add this to the model. The model extension points are 2: settings and options. Settings are project-wide, and probably apply to sanitizers, as the whole project should be built with the same sanitizing configuration
> 
> For extending the settings, the steps would be something like:
> 
> * Add the settings you want to `settings.yml` in your local cache (remember this file can be shared with `conan config install`)
> * Allow a value of `None` to this new setting. Something like `sanitizer = [None, address, thread]`
> * Add the `sanitizer` setting to your recipes: ``settings = "os", "compiler", ..., "sanitizer"
> * If you put sanitizer as a subsetting of the compiler, as probably they will be only defined for gcc or clang, then you don't need to add `sanitizer` to the recipe setting, as it will be `self.settings.compiler.sanitizer`
> * Use the `sanitizer` in your `build()` method, something like:
> 
> ```python
> def build(self):
>      cmake = CMake(self):
>      if self.settings.compiler.sanitizer == "address":
>            cmake.definitions["CONAN_CXX_FLAGS"] = "-fsanitize=address -fno-omit-frame-pointer"
>      elif self.settings.compiler.sanitizer == "thread":
>            ...
> ```

This approach would not work, as the current implementation of Conan maps the `CONAN_CXX_FLAGS` to `CMAKE_CXX_FLAGS` during the construction of the `CMake` build helper class:
https://github.com/conan-io/conan/blob/develop/conans/client/build/cmake.py#L75

This is not a good practice: business logic and behaviour in the constructor. It also hampers tweaking of these settings as per the example above and thereafter "running" Conan logic. And this results in clients having to mimic what Conan is doing internally.

I know, as I am dealing with this situation at this very moment.
I had to copy the logic from Conan's `CMakeDefinitionsBuilder.get_definitions()` method (https://github.com/conan-io/conan/blob/develop/conans/client/build/cmake_flags.py#L340) for the handling of the flag `_set_cmake_flags` as all this occurs via the constructor of the `CMake` build helper. This should be decoupled.

**Comment by hoyhoy** [comment]
> def build(self):
>      cmake = CMake(self):
>      if self.settings.compiler.sanitizer == "address":
>            cmake.definitions["CONAN_CXX_FLAGS"] = "-fsanitize=address -fno-omit-frame-pointer"
>      elif self.settings.compiler.sanitizer == "thread":

It's a nightmare because this works for the current package build but `-fsanitize=address` has to be applied across all dependencies for the binaries to link.  The only way I could figure out how to do this is apply the `{C,CXX,LD}FLAGS` in individual conan profiles, and even that doesn't work because `boost` uses `b2`.   I had to write custom `conanfile.py` `config_options()` to copy the `CFLAGS`, `CXXFLAGS` and `LDFLAGS` to `boost`'s `extra_b2_flags` for the sanitization flags to be applied.  I had huge issues with `protobuf` as well.

I have an incredible number of profiles to support this.  And the conan profile processor is quite terrible --  you can't include from an include, the variable replacement doesn't error if you make a typo, and so on...

It'd be a whole lot easier if I could modify the external environment (i.e. set `CFLAGS` on the dependencies) from the current package's `conanfile.py`, but it appears there's some sort of religious objection to that.

Ultimately, conan-ceneter should just support the sanitizer settings, but I can't see that happening.

As it is now, this is my current situation...

```
./fedora39/linux_x86_64_gcc13_tsan.profile
./fedora39/linux_x86_64_clang_latest_asan.profile
./fedora39/linux_x86_64_clang_latest_tsan.profile
./fedora39/linux_x86_64_gcc13_debug.profile
./fedora39/linux_x86_64_gcc13_asan.profile
./fedora39/linux_x86_64_gcc13_release.profile
./fedora39/linux_x86_64_clang_latest_debug.profile
./fedora39/linux_x86_64_clang_latest_msan.profile
./fedora38/linux_x86_64_gcc13_debug.profile
./fedora38/linux_x86_64_gcc13_asan.profile
./fedora38/linux_x86_64_gcc13_release.profile
./common/verbose.profile
./common/release.profile
./common/grpc.profile
./common/clear_sanitizer_options.profile
./common/ccache.profile
./common/gcc13.profile
./common/devtest.profile
./common/node_only.profile
./common/build_tools.profile
./common/macos14.profile
./common/clang_latest_rpm.profile
./common/gcc11.profile
./common/with_all.profile
./common/clang_defaults.profile
./common/xcode14.profile
./common/colorize.profile
./common/xcode15.profile
./common/macos13.profile
./common/clang_latest.profile
./common/cmake_local.profile
./common/debug.profile
./common/linux.profile
./arch/x86_64.profile
./macos14/macos_x86_64_xcode15_asan.profile
./macos14/macos_x86_64_xcode14_debug.profile
./macos14/macos_x86_64_xcode14_asan.profile
./macos14/macos_x86_64_clang_latest_asan.profile
./macos14/macos_x86_64_clang_latest_debug.profile
./macos14/macos_x86_64_xcode14_release.profile
./macos14/macos_x86_64_clang_latest_tsan.profile
./macos14/macos_x86_64_clang_latest_release.profile
./macos14/macos_x86_64_xcode15_debug.profile
./macos14/macos_x86_64_xcode15_release.profile
./macos13/macos_x86_64_xcode15_asan.profile
./macos13/macos_x86_64_xcode14_debug.profile
./macos13/macos_x86_64_xcode14_asan.profile
./macos13/macos_x86_64_clang_latest_asan.profile
./macos13/macos_x86_64_clang_latest_debug.profile
./macos13/macos_x86_64_xcode14_release.profile
./macos13/macos_x86_64_clang_latest_release.profile
./macos13/macos_x86_64_xcode15_debug.profile
./macos13/macos_x86_64_xcode15_release.profile
./rhel8/linux_x86_64_gcc11_release.profile
./rhel8/linux_x86_64_gcc13_tsan.profile
./rhel8/linux_x86_64_gcc11_asan.profile
./rhel8/linux_x86_64_clang_latest_asan.profile
./rhel8/linux_x86_64_gcc11_debug.profile
./rhel8/linux_x86_64_clang_latest_tsan.profile
./rhel8/gcc_toolset13.profile
./rhel8/linux_x86_64_gcc13_debug.profile
./rhel8/linux_x86_64_gcc13_asan.profile
./rhel8/linux_x86_64_gcc13_release.profile
./rhel8/linux_x86_64_clang_latest_debug.profile
./rhel8/linux_x86_64_clang_latest_msan.profile
./rhel8/linux_x86_64_gcc11_tsan.profile
./build/msan_ignore_list.txt
```

It'd be about 1000x easier to use Python to specify these compiler flags, but I can't.

**Comment by memsharded** [comment]
> I have an incredible number of profiles to support this. And the profile processor is quite terrible -- you can't include from an include, the variable replacement doesn't error if you make a typo, and so on...

Yes, you can include from an include. It is also possible now to use jinja templates which are very powerful too. If you have any specific feedback about that, it would be better to open a new ticket, I don't think this is really related to this one.

> cmake.definitions["CONAN_CXX_FLAGS"] 

This also seems legacy, referring to the old ``from conans import CMake`` and not the new ``from conan.tools.cmake import CMake`` helper (together with CMakeDeps and CMakeToolchain). Now we have ``confs`` like ``tools.build:cxxflags`` to be able to set flags not only to CMake, but to other build systems as well, at least our integrations for MSBuild, Meson, Autotools already take them into account

I think this ticket that dates back to 2021 is referring to legacy or outdated Conan versions and integrations. I think it is better to close it as outdated, and then file new tickets as necessary that take into accounts more updated Conan versions. Thanks for the feedback, looking forward it!

**Comment by hoyhoy** [comment]
> Yes, you can include from an include. 

You can, but the variable replacement doesn't work in 1.X

**Comment by memsharded** [comment]
> You can, but the variable replacement doesn't work.

Variable replacement has been superseded in Conan 1.X by jinja variables, and dropped in Conan 2.0

**Comment by hoyhoy** [comment]
How do you specify a `tools.build:cxxflags` per `build_type` and per `compiler.sanitizer`?

**Comment by memsharded** [comment]
You can put them in a ``flags_debug.profile`` and ``flags_release.profile`` that you either include or you pass them in command line like ``conan install --profile=windowsv143 --profile=flags_release.profile`` to compose them dynamically and avoid having to duplicate profiles

**Comment by hoyhoy** [comment]
So, a hundred separate conan profiles?  That's what I'm doing now.

Wouldn't something like this in a `conanfile.py`'s `config_options()` be easier?

```python
# DOES NOT WORK FOR DEPENDENCIES!!!!
if self.settings.build_type == "Debug":
    env = Environment()
    if self.compiler_name("gcc"):
        env.append("CXXFLAGS", "-0g -ggdb3")
        env.append("CFLAGS", "-0g -gggdb3")
    if self.compiler_name("clang"):
        env.append("CXXFLAGS", "-g -glldb -gcolumn-info")
        env.append("CFLAGS", "-g -glldb -gcolumn-info")
```

I noticed boost's conan-center `conanfile.py` uses a VirtualEnvironment, but it seems to be the only package using that.

**Comment by memsharded** [comment]
> So, a hundred separate conan profiles? That's what I'm doing now.

No, you can divide those many by using profile composition. Instead of having profiles for asan, just when you want those flags, add them as ``conan install ... -pr=windows.profile -pr=asan.profile`` or something like that. You can use envvars read by jinja so you can do ``CONAN_ASAN=1 conan install .... -pr=windows.profile`` (and profiles have conditionals on ``os.getenv()``). Or if you prefer to have one explicit different profile for each use case, then maybe generate all of them with some script?. There is also the ``profile.py`` plugin that you can use to inject/augment profiles (we are already doing it to deduce ``runtime_type`` based on ``build_type`` for example). So there are multiple ways to reduce the number of needed profiles.

> Wouldn't something like this in a conanfile.py's config_options() be easier?

No, that is order of magnitude more complex and problematic. The moment you can define configuration of your dependencies directly in conanfile, you open a whole can of worms of conflicts, priorities, composition, etc. We have already suffered it with options for a long time, no way we are extending that nightmare to compiler flags.

**Comment by hoyhoy** [comment]
> The moment you can define configuration of your dependencies directly in conanfile...

The boost package in conan-center extended compiler flags to options via `extra_b2_flags` which you can set via `config_options`.  The only reason I had to configure boost locally was because the boost build only observed `tools.build:cxxflags` and `[buildenv]` not the regular environment set via the conan profile.

I originally thought something was wrong with boost's conan-io package, but that's just how it's built.  The conan profile's `[env]` section is NOT observed.  This is the only package doing it like this.

https://github.com/conan-io/conan-center-index/issues/21604

I could have used `tools.build:cxxflags`, but there's not a good way to append to Python lists using profiles with conan 1.X.  Is this fixed in 2.x?  I managed to make it work by appending strings to construct Python lists, but that seemed like not a best practice either.

From my perspective, it seems kinda crazy to not just use Python to do all of this as you can't do simple things like `if build_type == "Debug"` using the profile interpreter.

**Comment by memsharded** [comment]
> The boost package in conan-center extended compiler flags to options via extra_b2_flags which you can set via config_options. The only reason I had to configure boost locally was because the boost build only observed tools.build:cxxflags and [buildenv] not the regular environment set via the conan profile.

It can inject extra flags for itself, not for dependencies. The problems come when trying to set compiler flags for dependencies. CMakeToolchain even has a way to directly inject flags from recipes into itself, check https://docs.conan.io/2/reference/tools/cmake/cmaketoolchain.html#extra-compilation-flags. But it is for the current package, not for dependencies.

> I could have used tools.build:cxxflags, but there's not a good way to append to Python lists using profiles with conan 1.X. Is this fixed in 2.x?

All ``confs`` items, including ``tools.build:cxxflags`` can be defined, appended, pre-pended and unset, with syntax like ``tools.build:cxxflags+=['cxxflags']``. Also in Conan 1.X I think.


> From my perspective, it seems kinda crazy to not just use Python to do all of this as you can't do simple things like if build_type == "Debug" using the profile interpreter.

For generic behavior, you can do it in python in the ``profile.py`` plugin.

**Comment by hoyhoy** [comment]
Yeah, I could make a string that turned into an array with profile variables, but using multiple arrays didn't work with 1.X.

**Comment by memsharded** [comment]
Yes, that works in Conan 1.X. This test is passing in Conan 1.X. As you can see, it is correctly appending lists for ``tools.build:cxxflags``:

```python
def test_basic_composition():
    client = TestClient()
    conanfile = textwrap.dedent("""
           from conans import ConanFile

           class Pkg(ConanFile):
               def generate(self):
                   for k, v in self.conf.items():
                       self.output.info("{}${}".format(k, v))
           """)
    client.save({"conanfile.py": conanfile})
    profile1 = textwrap.dedent("""\
        [conf]

        tools.build:cxxflags=["-flag1", "-flag2"]
        """)
    profile2 = textwrap.dedent("""\
        [conf]
        tools.build:cxxflags+=["-flag3", "-flag4"]
        """)
    client.save({"profile1": profile1,
                 "profile2": profile2})
    client.run("install . -pr=profile1 -pr=profile2")
    assert "tools.build:cxxflags=['-flag1', '-flag2', '-flag3', '-flag4']" in client.out
    assert "conanfile.py: tools.build:cxxflags$['-flag1', '-flag2', '-flag3', '-flag4']" in client.out
```

**Comment by hoyhoy** [comment]
Yeah it doesn't work with 1.X using conan.profiles.  You have to make the string look like an array for it to work with variable replacement.

```
#global.conf
tools.build:cflags=["-glldb", "-fdiagnostics-show-option", "-fcaret-diagnostics"]
tools.build:cxxflags=["-glldb", "-fdiagnostics-show-option", "-fcaret-diagnostics"]
```

```
#conan.profile...

DISABLED_WARNINGS=["-Wno-unused-but-set-variable", "-Wno-deprecated-declarations", "-Wno-unused-command-line-argument", 
                   "-Wno-null-pointer-subtraction", "-Wno-deprecated-builtins", "-Wno-enum-constexpr-conversion", 
                   "-Wno-implicit-function-declaration", "-Wno-sizeof-array-div", "-Wno-incompatible-function-pointer-types",
                   "-Wno-int-conversion"]
CLANG_COMPILER_FLAGS=["-fno-common", "-fno-strict-aliasing", "-fno-omit-frame-pointer", "-fno-optimize-sibling-calls"]
ASAN_COMPILER_FLAGS=["-fPIE", "-fsanitize-address-use-after-scope", "-fsanitize=address" "-fsanitize-address-use-after-return=always" ]
CLANG_DEBUG_FLAGS=["-g", "-glldb", "-gcolumn-info"]

[conf]
tools.build:cxxflags = $CLANG_COMPILER_FLAGS + $DISABLED_WARNINGS + $CLANG_DEBUG_FLAGS
tools.build:cflags = $CLANG_COMPILER_FLAGS + $DISABLED_WARNINGS + $CLANG_DEBUG_FLAGS

app:tools.build:cxxflags = $CLANG_COMPILER_FLAGS
app:tools.build:cflags = $CLANG_COMPILER_FLAGS
```

Build failure...
```
CFLAGS=["-fno-common","-fno-strict-aliasing","-fno-omit-frame-pointer","-fno-optimize-sibling-calls"] ["-g","-glldb","-gcolumn-info"] ["-fPIE","-fsanitize-address-use-after-scope","-fsanitize=address" "-fsanitize-address-use-after-return=always" ] ["-Wno-unused-but-set-variable","-Wno-deprecated-declarations","-Wno-unused-command-line-argument","-Wno-null-pointer-subtraction","-Wno-deprecated-builtins","-Wno-enum-constexpr-conversion","-Wno-implicit-function-declaration","-Wno-sizeof-array-div","-Wno-incompatible-function-pointer-types","-Wno-int-conversion"]

  Building C object CMakeFiles/cmTC_83579.dir/testCCompiler.c.o
    /usr/local/bin/clang   "-fno-common":"-fno-strict-aliasing":"-fno-omit-frame-pointer":"-fno-optimize-sibling-calls"] ["-g":"-glldb":"-gcolumn-info"] ["-fPIE":"-fsanitize-address-use-after-scope":"-fsanitize=address" "-fsanitize-address-use-after-return=always" ] ["-Wno-unused-but-set-variable":"-Wno-deprecated-declarations":"-Wno-unused-command-line-argument":"-Wno-null-pointer-subtraction":"-Wno-deprecated-builtins":"-Wno-enum-constexpr-conversion":"-Wno-implicit-function-declaration":"-Wno-sizeof-array-div":"-Wno-incompatible-function-pointer-types":"-Wno-int-conversion":-O2 -I/usr/include     -m64 -glldb -fdiagnostics-show-option -fcaret-diagnostics  -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk -mmacosx-version-min=14.1 -fPIE -fcolor-diagnostics -MD -MT CMakeFiles/cmTC_83579.dir/testCCompiler.c.o -MF CMakeFiles/cmTC_83579.dir/testCCompiler.c.o.d -o CMakeFiles/cmTC_83579.dir/testCCompiler.c.o -c /Users/jperrie/.conan/data/lua/5.4.4/_/_/build/fae3d9242d93235df50cad4542f276fd3795da01/build/Debug/CMakeFiles/CMakeScratch/TryCompile-KokZME/testCCompiler.c
```

**Comment by hoyhoy** [comment]
The other issue I was having was if specified `[conf]` in two different include files, the first one was the only one applied.

**Comment by hoyhoy** [comment]
This doesn't work either...

```
# conan.profile
DISABLED_WARNINGS=["-Wno-unused-but-set-variable", "-Wno-deprecated-declarations", "-Wno-unused-command-line-argument", 
                   "-Wno-null-pointer-subtraction", "-Wno-deprecated-builtins", "-Wno-enum-constexpr-conversion", 
                   "-Wno-implicit-function-declaration", "-Wno-sizeof-array-div", "-Wno-incompatible-function-pointer-types",
                   "-Wno-int-conversion"]
CLANG_COMPILER_FLAGS=["-fno-common", "-fno-strict-aliasing", "-fno-omit-frame-pointer", "-fno-optimize-sibling-calls"]
ASAN_COMPILER_FLAGS=["-fPIE", "-fsanitize-address-use-after-scope", "-fsanitize=address" "-fsanitize-address-use-after-return=always" ]
CLANG_DEBUG_FLAGS=["-g", "-glldb", "-gcolumn-info"]

[conf]
tools.build:cxxflags += $CLANG_COMPILER_FLAGS
tools.build:cxxflags += $DISABLED_WARNINGS
tools.build:cxxflags += $CLANG_DEBUG_FLAGS
tools.build:cflags += $CLANG_COMPILER_FLAGS 
tools.build:cflags += $DISABLED_WARNINGS 
tools.build:cflags += $CLANG_DEBUG_FLAGS

app:tools.build:cxxflags += $CLANG_COMPILER_FLAGS
app:tools.build:cflags += $CLANG_COMPILER_FLAGS
```

**Comment by hoyhoy** [comment]
This is why I was saying relying on profiles to do this makes everything really difficult.  There are Python profiles in conan 2.X?  We're working on upgrading, but we have a bunch old C++ dependencies on cloud libraries, and other dependencies like redis-server that don't exist in conan-center.

**Comment by hoyhoy** [comment]
The other issue is that conan profiles aren't run from the repository during build.  They're deployed to `$USER/.conan`.  This makes it even harder to fix our build system because then we have to deploy new profiles.  That's caused a number of issues for us.

**Comment by memsharded** [comment]
Thanks for the feedback and the detailed examples, those help to try to understand the issue.
It is true that the first syntax ``tools.build:cxxflags = $FLAGS1 + $FLAGS2``, doesn't work because python list addition is not possible with variables, those are not python variables. But you can perfectly do:
```
tools.build:cxxflags+= $FLAGS1
tools.build:cxxflags+= $FLAGS2
```
And it achieves the desired result without problems.

I am trying to reproduce it with another different unit test, and it seems it is working, I cannot make it fail.
If you could please have a look to it and check what could be the gap, that would be great:

```python

def test_basic_composition_global_conf():
    client = TestClient()
    conanfile = textwrap.dedent("""
       from conans import ConanFile

       class Pkg(ConanFile):
           def generate(self):
               for k, v in self.conf.items():
                   self.output.info("{}${}".format(k, v))
       """)
    client.save({"conanfile.py": conanfile})
    global_conf = textwrap.dedent("""\
        tools.build:cxxflags=["-flag1", "-flag2"]
        """)
    save(client.cache.new_config_path, global_conf)
    profile2 = textwrap.dedent("""\
        FLAGS1 = ["-flag3", "-flag4"]
        FLAGS2 = ["-flag5", "-flag6"]
        [conf]
        tools.build:cxxflags+= $FLAGS1
        tools.build:cxxflags+= $FLAGS2
        """)
    client.save({"profile2": profile2})
    client.run("install . -pr=profile2")
    assert "tools.build:cxxflags=['-flag1', '-flag2', '-flag3', '-flag4', '-flag5', '-flag6']" in client.out
    assert "conanfile.py: tools.build:cxxflags$['-flag1', '-flag2', '-flag3', '-flag4', '-flag5', '-flag6']" in client.out
```

**Comment by hoyhoy** [comment]
Probably there's some `repr()` call in between the actual profiles and conan.  The `+=` doesn't work in 1.X if you put it in an actual profile.  Conan emits the flags as `CFLAGS=["-fno-common","-fno-strict-aliasing","-fno-omit-frame-pointer","-fno-optimize-sibling-calls"] ["-g","-glldb","-gcolumn-info"] ` (space delimited).
